contracts/governance/src/governance.rs

```diff
@@ -0,0 +1,274 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+use core::iter;
+
+use alloc::vec;
+use alloc::vec::Vec;
+use canonical_derive::Canon;
+use dusk_bls12_381::BlsScalar;
+use dusk_hamt::Map;
+use dusk_jubjub::GENERATOR_EXTENDED;
+use dusk_pki::PublicKey;
+use dusk_schnorr::Signature;
+
+use crate::*;
+
+#[derive(Debug, Default, Clone, Canon)]
+pub struct GovernanceContract {
+    pub(crate) seeds: Map<BlsScalar, ()>,
+    pub(crate) balances: Map<PublicKey, u64>,
+    pub(crate) whitelist: Map<PublicKey, ()>,
+    pub(crate) paused: bool,
+    pub(crate) total_supply: u64,
+}
+
+impl GovernanceContract {
+    /// Authority of the contract
+    ///
+    /// Will have to be defined in the constant space so the bytecode of the
+    /// contract will be changed as the authority does
+    pub const AUTHORITY: PublicKey =
+        PublicKey::from_raw_unchecked(GENERATOR_EXTENDED);
+
+    fn validate_seed(
+        &mut self,
+        arguments: Vec<BlsScalar>,
+        signature: Signature,
+    ) -> Result<(), Error> {
+        let seed = arguments[0];
+
+        if self.seeds.get(&seed)?.is_some() {
+            return Err(Error::SeedAlreadyUsed);
+        }
+
+        #[cfg(target_arch = "wasm32")]
+        if !rusk_abi::verify_schnorr_sign(
+            signature,
+            Self::AUTHORITY,
+            rusk_abi::poseidon_hash(arguments),
+        ) {
+            return Err(Error::InvalidSignature);
+        }
+
+        #[cfg(not(target_arch = "wasm32"))]
+        if !signature
+            .verify(&Self::AUTHORITY, dusk_poseidon::sponge::hash(&arguments))
+        {
+            return Err(Error::InvalidSignature);
+        }
+
+        self.seeds.insert(seed, ())?;
+
+        Ok(())
+    }
+
+    fn check_pause(&self) -> Result<(), Error> {
+        (!self.paused).then_some(()).ok_or(Error::ContractIsPaused)
+    }
+
+    fn is_allowed(&self, address: &PublicKey) -> Result<(), Error> {
+        self.whitelist
+            .get(address)?
+            .as_deref()
+            .copied()
```

Comment:
As far as I know, `get` will return a branch of the `whitelist` map to the `address`. Why is it then necessary to also run `as_deref().copied()`?

Reply:
It is necessary once you change the feature. It is not required for the vector feature. So I think we should suppress that warning since it allows seamless integration.

Reply:
This will be taken care of once #775 merges. We decided to switch to having a [vector](https://github.com/dusk-network/rusk/pull/775/files#diff-a77103439ec63a2394383bd42d8f9361383b22c8d195a52a8f451471e16f6d61R13)  instead of `dusk_hamt::Map` because it requires less gas fees to run which could be tested at contracts/governance /tests

In this branch:

https://github.com/dusk-network/rusk/blob/governance/gas_test_final/contracts/governance/tests/gas_fees.rs#L50)